docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+utilising the [SSH agent protocol][ssh-agent].
+
+On macOS, the distribution `ssh-agent` integrates with the system keychain.
+Other choices include the `ssh-agent` provided by the `openssh` suite, the
+`gpg-agent`, or (on Linux systems) `gnome-keyring`.
+
+Lastly, we consider [D-Bus][dbus] for publish-subscribe messaging. D-Bus is
+available on `systemd`-based systems per default, while it requires use of an
+external package manager on macOS.
+
+> Note: a D-Bus implementation is provided on `systemd`-based platforms, and can
+> be installed on macOS using `homebrew` or `MacPorts`. It may, however, turn
+> out to complicate builds due to varying link-time requirements. As we only
+> intend to use D-Bus for pub-sub messaging ("signals"), we may instead consider
+> to provide a purpose-built in-memory pub-sub service.
+
+
+### Signatures
+
+Each process which needs to produce signatures using the `radicle-link` _device
+key_, SHALL delegate this to an agent conforming to the [SSH agent
+protocol][ssh-agent] (section 4.5.).
+
+Key access restrictions thus depend on user configuration: the agent could be
+configured to prompt for a passphrase after a timeout elapses (or never),
+require presence of a hardware security token, or utilise platform
+authenticators such as TouchID&trade;.
+
+Key generation SHOULD, however, be performed using `radicle-link`-supplied
+tooling building on the [ed25519-zebra] library, which prevents certain weaker
+keys from being generated.
+
+Signature verification MUST be performed using the [ed25519-zebra] library, and
+thus MUST NOT be delegated to the agent.
+
+Note that (hypothetical) applications which require a large number of signatures
+to be generated (such as data migration tools) may find `ssh-agent` throughput
+insufficient, and require the raw key material to be supplied directly. This
+means that long-term secure key storage remains a concern of the `radicle-link`
+application suite.
+
+
+### Process Orchestration
+
+All daemon processes SHALL be started on-demand using the platform socket
+activation protocol. Unless otherwise noted, daemons SHALL terminate after a
+configured period of idle time (in the order of 10s of minutes). The platform
+process manager MUST NOT restart socket-activated daemon processes if they exit
+with a success status (except by another request for the socket).
+
+All daemon processes SHALL run with user privileges (typically the logged-in
+user). Further confinement (e.g. SELinux) is beyond the scope of this document.
+
+
+### IPC
+
+All communication with daemon processes SHALL occur over UNIX domain sockets in
+`SOCK_STREAM` mode. The socket files MUST be stored in a directory only
+accessible by the logged-in user (typically `$XDG_RUNTIME_DIR`), and have
+permissions `0700` by default. Per-service socket paths are considered "well
+known".
+
+RPC calls over those sockets use [CBOR] for their payload encoding. As
```

Comment:
Am I right in thinking that this RFC is not specifying what we're going to use as an RPC?

Reply:
No, it is pretty explicit: you send `<u32><CBOR payload>`

Reply:
No, it is pretty explicit: you send `<u32><CBOR payload>`

Reply:
And get back `<u32><CBOR payload>`.

Reply:
What `CBOR payload` is is given in IPC. What `payload` of `CBOR payload` is, may be defined in #673, or a subsequent document.

Reply:
Right, didn't put 2 and 2 together. So that's the message format and the transport is the socket.